<?xml version="1.0" encoding="UTF-8"?>
<mule xmlns="http://www.mulesoft.org/schema/mule/core"
      xmlns:test="http://www.mulesoft.org/schema/mule/test"
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="
    http://www.mulesoft.org/schema/mule/core http://www.mulesoft.org/schema/mule/core/current/mule.xsd
    http://www.mulesoft.org/schema/mule/test http://www.mulesoft.org/schema/mule/test/current/mule-test.xsd">

   <flow name="tryInErrorHandler">
        <raise-error type="APP:SOME"/>
        <error-handler>
            <on-error-continue>
                <try>
                    <set-payload value="hello"/>
                    <error-handler>
                        <on-error-continue>
                            <test:invocation-counter name="try-in-eh"/>
                        </on-error-continue>
                    </error-handler>
                </try>
            </on-error-continue>
        </error-handler>
    </flow>

    <flow name="tryWithRecursiveOnErrorContinueInsideSubflow">
        <flow-ref name="tryWithRecursiveOnErrorContinueSubflow"/>
        <error-handler >
            <on-error-continue type="CUSTOM:CUSTOM_ERROR">
                <logger level="INFO" message="Main flow strategy reached." />
                <test:queue name="dlq"/>
            </on-error-continue>
        </error-handler>
    </flow>

    <flow name="tryWithRecursiveOnErrorContinueInsideFlow">
        <flow-ref name="tryWithRecursiveOnErrorContinueFlow"/>
        <error-handler >
            <on-error-continue type="CUSTOM:CUSTOM_ERROR">
                <logger level="INFO" message="Main flow strategy reached." />
                <test:queue name="dlq"/>
            </on-error-continue>
        </error-handler>
    </flow>

    <sub-flow name="tryWithRecursiveOnErrorContinueSubflow">
        <try>
            <raise-error type="CUSTOM:CUSTOM_ERROR" description="Custom Error is Raised" />
            <error-handler >
                <on-error-continue type="CUSTOM:CUSTOM_ERROR" when="(vars.errorCallCount default 0) &lt; 2">
                    <set-variable value="#[(vars.errorCallCount default 0) + 1]" variableName="errorCallCount"/>
                    <logger level="INFO" message="Retrying subflow processing." />
                    <flow-ref name="tryWithRecursiveOnErrorContinueSubflow" />
                </on-error-continue>
                <on-error-propagate type="CUSTOM:CUSTOM_ERROR">
                    <logger level="INFO" message="Retries exhausted. Propagating error." />
                </on-error-propagate>
            </error-handler>
        </try>
    </sub-flow>

    <flow name="tryWithRecursiveOnErrorContinueFlow">
        <try>
            <raise-error type="CUSTOM:CUSTOM_ERROR" description="Custom Error is Raised" />
            <error-handler >
                <on-error-continue type="CUSTOM:CUSTOM_ERROR" when="(vars.errorCallCount default 0) &lt; 2">
                    <set-variable value="#[(vars.errorCallCount default 0) + 1]" variableName="errorCallCount"/>
                    <logger level="INFO" message="Retrying subflow processing." />
                    <flow-ref name="tryWithRecursiveOnErrorContinueSubflow" />
                </on-error-continue>
                <on-error-propagate type="CUSTOM:CUSTOM_ERROR">
                    <logger level="INFO" message="Retries exhausted. Propagating error." />
                </on-error-propagate>
            </error-handler>
        </try>
    </flow>

</mule>